pythonPackages.cot: init at 2.2.1

evanjs opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change
> [COT ](https://github.com/glennmatthews/cot) (the Common OVF Tool) is a tool for editing Open Virtualization Format (.ovf, .ova) virtual appliances, with a focus on virtualized network appliances such as the Cisco CSR 1000V and Cisco IOS XRv platforms.

I figured this might be more accessible/helpful than e.g. https://code.vmware.com/tool/ovf

This might prove useful in adding [ESXi virtualization support](https://github.com/NixOS/nixpkgs/issues/94104) ([Discourse post](https://discourse.nixos.org/t/virtualization-ova-ovf-esxi-support-extended-ova-customization/7536)) if/once https://github.com/glennmatthews/cot/issues/80 is resolved.

Until then, I think this might still be useful for managing OVF appliances, either during or after their generation, via the `virtualbox-image` and related expressions (`VBoxManage export`, etc).
___
###### Things done
- Added `cot` to python-modules (>=python3)
- Added `cot` to `top-level/pythonPackages`
- Added `cot` to `top-level/all-packages` (via `toPythonApplication`)
___
##### Many tests are currently disabled
Without any filters, this is the output from a normal build (with all tests enabled):
```
python3.8-cot> =========== 189 failed, 380 passed, 24 warnings, 17 errors in 25.32s ===========
builder for '/nix/store/lakc1ns9f44l1hxw778207danhr37fim-python3.8-cot-2.2.0.drv' failed with exit code 1; last 10 log lines:
  ERROR COT/ui/tests/test_cli.py::TestCLIAddDisk::test_invalid_args - Assertion...
  ERROR COT/ui/tests/test_cli.py::TestCLIAddDisk::test_nonexistent_file - Asser...
  ERROR COT/ui/tests/test_cli.py::TestCLIEditHardware::test_args_append - Asser...
  ERROR COT/ui/tests/test_cli.py::TestCLIEditProperties::test_set_property_valid
  ERROR COT/vm_description/ovf/tests/test_item.py::TestOVFItem::test_set_property
  ERROR COT/vm_description/ovf/tests/test_item.py::TestOVFItem::test_set_property_no_modification
  ERROR COT/vm_description/ovf/tests/test_ovf.py::TestOVFInputOutput::test_input_output_bad_file
  ERROR COT/vm_description/ovf/tests/test_ovf.py::TestOVFInputOutput::test_tar_links
  ERROR COT/vm_description/ovf/tests/test_ovf.py::TestOVFInputOutput::test_tar_untar
  =========== 189 failed, 380 passed, 24 warnings, 17 errors in 25.32s ===========
[0 built (1 failed), 0.0 MiB DL]
error: build of '/nix/store/lakc1ns9f44l1hxw778207danhr37fim-python3.8-cot-2.2.0.drv' failed
```

I ignored any files with failing tests, vs disabling each test that was failing.
This results in 19 lines of `--ignore` vs ~189.

Note that this package needs `py3k`  and up, and is disabled `if !isPy3k` in the expression.
I noticed that some packages do this checking earlier (e.g. in `python-packages.nix`), and while I'm not sure which is a best practice, the `disabled` attribute does prevent `cot` from being built with `python2` (`error: cot-2.2.1 not supported for interpreter python2.7`)

evanjs committed:
pythonPackages.cot: init at 2.2.1

sikmir commented on pkgs/development/python-modules/cot/default.nix:
```diff
@@ -0,0 +1,66 @@
+
+    # ignore files containing failing tests
+    pytest \
+      --ignore=COT/commands/tests/test_add_disk.py \
```
Try to use `disabledTests` instead.

evanjs replied:
Ah, did not see that, though I did see some custom implementations (e.g. `skipTests`, etc)
I will see if this works.

jonringer replied:
disabledTests only works with `pytestCheckHook`

evanjs replied:
@jonringer Yeah, I noticed that at some point, and then found the hooks needed for the other libraries, but I wasn't able to determine what I needed to input to get `disabledTests` working properly.

Just using `pytestCheckHook` and `pytestFlagsArray` looks a bit nicer, anyhow.
I'm sure it would be even better if I got the `disabledTests` arguments right... 

<details>
<summary><span><code>pytestCheckHook</span></code> and <span><code>pytestFlagsArray</code></span></summary>

``` diff
diff --git a/pkgs/development/python-modules/cot/default.nix b/pkgs/development/python-modules/cot/default.nix
index f6baef515f4..4a5da3204a4 100644
--- a/pkgs/development/python-modules/cot/default.nix
+++ b/pkgs/development/python-modules/cot/default.nix
@@ -1,7 +1,7 @@
 { stdenv, buildPythonPackage, fetchPypi, pythonOlder, isPy3k
 , argcomplete, colorlog, pyvmomi, requests, verboselogs
 , psutil, pyopenssl, setuptools
-, mock, pytest
+, mock, pytest-mock, pytestCheckHook
 }:

 buildPythonPackage rec {
@@ -18,35 +18,28 @@ buildPythonPackage rec {
   propagatedBuildInputs = [ colorlog pyvmomi requests verboselogs pyopenssl setuptools ]
   ++ stdenv.lib.optional (pythonOlder "3.3") psutil;

-  checkInputs = [ mock pytest ];
-
-  checkPhase = ''
-    runHook preCheck
-
-    # ignore files containing failing tests
-    pytest \
-      --ignore=COT/commands/tests/test_add_disk.py \
-      --ignore=COT/commands/tests/test_add_file.py \
-      --ignore=COT/commands/tests/test_edit_hardware.py \
-      --ignore=COT/commands/tests/test_edit_product.py \
-      --ignore=COT/commands/tests/test_edit_properties.py \
-      --ignore=COT/commands/tests/test_inject_config.py \
-      --ignore=COT/commands/tests/test_install_helpers.py \
-      --ignore=COT/commands/tests/test_remove_file.py \
-      --ignore=COT/disks/tests/test_disk_representation.py \
-      --ignore=COT/disks/tests/test_iso.py \
-      --ignore=COT/disks/tests/test_qcow2.py \
-      --ignore=COT/disks/tests/test_raw.py \
-      --ignore=COT/disks/tests/test_vmdk.py \
-      --ignore=COT/helpers/tests/test_helper.py \
-      --ignore=COT/helpers/tests/test_qemu_img.py \
-      --ignore=COT/tests/test_utilities.py \
-      --ignore=COT/ui/tests/test_cli.py \
-      --ignore=COT/vm_description/ovf/tests/test_item.py \
-      --ignore=COT/vm_description/ovf/tests/test_ovf.py
-    runHook postCheck
-  '';
-
+  checkInputs = [ mock pytest-mock pytestCheckHook ];
+  pytestFlagsArray = [
+     "--ignore=COT/commands/tests/test_add_disk.py"
+     "--ignore=COT/commands/tests/test_add_file.py"
+     "--ignore=COT/commands/tests/test_edit_hardware.py"
+     "--ignore=COT/commands/tests/test_edit_product.py"
+     "--ignore=COT/commands/tests/test_edit_properties.py"
+     "--ignore=COT/commands/tests/test_inject_config.py"
+     "--ignore=COT/commands/tests/test_install_helpers.py"
+     "--ignore=COT/commands/tests/test_remove_file.py"
+     "--ignore=COT/disks/tests/test_disk_representation.py"
+     "--ignore=COT/disks/tests/test_iso.py"
+     "--ignore=COT/disks/tests/test_qcow2.py"
+     "--ignore=COT/disks/tests/test_raw.py"
+     "--ignore=COT/disks/tests/test_vmdk.py"
+     "--ignore=COT/helpers/tests/test_helper.py"
+     "--ignore=COT/helpers/tests/test_qemu_img.py"
+     "--ignore=COT/tests/test_utilities.py"
+     "--ignore=COT/ui/tests/test_cli.py"
+     "--ignore=COT/vm_description/ovf/tests/test_item.py"
+     "--ignore=COT/vm_description/ovf/tests/test_ovf.py"
+    ];

   # argparse is part of the standardlib
   prePatch = ''
```

</details>

jonringer replied:
disabledTests relate to filtering the test names. 

You should be able to use bash expansion as well, but I haven't used this with pytestFlagsArray:
```
"--ignore=COT/disks/tests/test_{vmdk,qcow2,raw,iso}.py"
```

jonringer replied:
I'm also not a big fan of, "disable failing tests". Generally we should be only disabling tests that don't work well in a sandbox. Test failures maybe indicative that the package isn't fully usable.

evanjs replied:
Yeah, many of the tests seem to use `ovftool`, and others seem to require network connectivity.
At some point, I would like to get `ovftool` added to nixpkgs, anyway, but for now, this is the list I have for `disabledTests`:
``` nix
  disabledTests = [
    "HelperGenericTest"
    "TestCOTAddDisk"
    "TestCOTAddFile"
    "TestCOTEditHardware"
    "TestCOTEditProduct"
    "TestCOTEditProperties"
    "TestCOTInjectConfig"
    "TestISO"
    "TestOVFAPI"
    "TestQCOW2"
    "TestRAW"
    "TestVMDKConversion"
  ];
```

```
=============== 367 passed, 210 deselected, 3 warnings in 5.61s ================
```

jonringer replied:
Do you mind add that as a comment? :)

sikmir commented on pkgs/development/python-modules/cot/default.nix:
```diff
@@ -0,0 +1,66 @@
+    '';
+    homepage = "https://github.com/glennmatthews/cot";
+    license = licenses.mit;
+    maintainers = [ "evanjs" ];
```
```suggestion
    maintainers = with maintainers; [ evanjs ];
```

jonringer commented:
please squash the maintainers commit

evanjs commented:
@jonringer I added a note to the `disabledTests` re ovftool and squashed the maintainers commit
I also removed the `checkPhase` so now only `disabledTests` is specified, making it a bit cleaner.

Anything else you can think of?

evanjs commented:
@ofborg build cot

jonringer reviewed (approved):
LGTM

Result of nixpkgs-review pr 94179 [1](https://github.com/Mic92/nixpkgs-review)<details>  <summary>2 packages built:</summary>  - cot (python38Packages.cot)  - python37Packages.cot</details>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/virtualization-ova-ovf-esxi-support-extended-ova-customization/7536/2
